### PR TITLE
soef connection refactoring. test coverage.

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

soef connection async friendly, better tests and full coverage.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/fetchai/connections/soef/connection.py [hunk 1]
```diff
@@ -56,9 +56,57 @@
 PUBLIC_ID = PublicId.from_str("fetchai/soef:0.3.0")
 
 
+NOT_SPECIFIED = object()
+
+PERSONALITY_PIECES = [
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:61]: Better to call it `PERSONALITY_PIECES_KEYS` imo

File: packages/fetchai/connections/soef/connection.py [hunk 2]
```diff
@@ -77,7 +126,14 @@ def __init__(
         :param soef_port: the SOEF port.
         :param excluded_protocols: the protocol ids excluded
         :param restricted_to_protocols: the protocol ids restricted to
+        :param chain_identifier: supported chain id
         """
+        if (
+            chain_identifier is not None
+            and chain_identifier not in self.SUPPORTED_CHAIN_IDENTIFIERS
+        ):
+            raise ValueError("Unsupported chain_identifier")
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:135]: Print the list of supported identifiers in the error message please

File: packages/fetchai/connections/soef/connection.py [hunk 3]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
         self.in_queue = None  # type: Optional[asyncio.Queue]
+        self._executor_pool: Optional[ThreadPoolExecutor] = None
+        self.chain_identifier: str = chain_identifier or "fetchai"
 
-    def send(self, envelope: Envelope) -> None:
+    @staticmethod
+    def _is_compatible_query(query: Query) -> bool:
         """
-        Send message handler.
+        Check if a query is compatible with the soef.
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:156]: Add a comment in new line: "Each query must contain a distance constraint type."

File: packages/fetchai/connections/soef/connection.py [hunk 4]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
+        :return: bool
+        """
+        filters = []
+
+        for constraint in equality_constraints:
+            if constraint.attribute_name not in PERSONALITY_PIECES:
+                continue
+            filters.append(
+                constraint.attribute_name + "," + constraint.constraint_type.value
+            )
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:189]: Should we not remove the constraint from the list (as it cannot be a service key if it is a personality piece!)?

Comment (DavidMinarsch) [reply packages/fetchai/connections/soef/connection.py:189]: I see below; you check twice for personality pieces; ignore above

File: packages/fetchai/connections/soef/connection.py [hunk 5]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
+        if not filters:  # pragma: nocover
+            return {}
+        return {"ppfilter": filters}
+
+    @staticmethod
+    def _construct_service_key_filter_params(
+        equality_constraints: List[Constraint],
+    ) -> Dict[str, List[str]]:
+        """
+        Construct a dictionary of service keys filters.
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:199]: Add a line: `We assume each equality constraint which is not a personality piece relates to a service key!"

File: packages/fetchai/connections/soef/connection.py [hunk 6]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
-                str(self.search_id),
+        err_ops = OefSearchMessage.OefErrorOperation
+
+        oef_error_operation = err_ops.OTHER
+
+        try:
+            if self.unique_page_address is None:  # pragma: nocover
+                await self._register_agent()
+
+            assert isinstance(envelope.message, OefSearchMessage), ValueError(
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:271]: should be before the line 268

File: packages/fetchai/connections/soef/connection.py [hunk 7]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
+        """
+        Send message handler.
+
+        :param envelope: the envelope.
+        :return: None
+        """
+        self._check_protocol_valid(envelope)
+        await self.process_envelope(envelope)
+
+    async def _request_text(self, *args, **kwargs) -> str:
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:248]: Where is this used?

File: packages/fetchai/connections/soef/connection.py [hunk 8]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
-                "Service description incompatible with SOEF: values={}".format(
-                    service_description.values
-                )
+        :param oef_message: OefSearchMessage
+        """
+        service_description = oef_message.service_description
+
+        data_model_handlers = {
+            "location_agent": self._register_location_handler,
+            "personality_agent": self._register_personality_piece_handler,
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:315]: I think it should be called `set_personality_piece`

File: packages/fetchai/connections/soef/connection.py [hunk 9]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
+            "set_service_key": self._set_service_key_handler,
+            "remove_service_key": self._remove_service_key_handler,
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:317]: Should there not be a `remove_personality_piece` as well?

File: packages/fetchai/connections/soef/connection.py [hunk 10]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
-        :return: bool
+        :param service_description: Service description
+        :return None
         """
-        is_compatible = (
-            isinstance(service_description.values.get("location", None), Location)
-        ) or (
-            isinstance(service_description.values.get("piece", None), str)
-            and isinstance(service_description.values.get("value", None), str)
+        self._sure_data_model(service_description, "set_service_key")
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:336]: Sure? Typo?

File: packages/fetchai/connections/soef/connection.py [hunk 11]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
+    async def _register_location_handler(
+        self, service_description: Description
+    ) -> None:
+        """
+        Register service with location.
+
+        :param service_description: Service description
+        :return None
+        """
+        self._sure_data_model(service_description, "location_agent")
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:420]: magic values like `location_agent` and `remove_service_key` should be specified as reusable constants!

File: packages/fetchai/connections/soef/connection.py [hunk 12]
```diff
@@ -91,106 +147,349 @@ def __init__(
...
+
+        agent_location = service_description.values.get("location", None)
+        if agent_location is None or not isinstance(
+            agent_location, Location
+        ):  # pragma: nocover
+            raise SOEFException.debug("Bad location provided.")
+        await self._set_location(agent_location)
+
+    def _sure_data_model(
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:429]: Naming issue

File: packages/fetchai/connections/soef/connection.py [hunk 13]
```diff
@@ -200,48 +499,37 @@ def _register_agent(self) -> None:
         url = parse.urljoin(self.base_url, "register")
         params = {
             "api_key": self.api_key,
-            "chain_identifier": "fetchai",
+            "chain_identifier": self.chain_identifier,
             "address": self.address,
             "declared_name": self.declared_name,
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/soef/connection.py:504]: We should also add the new `architecture` identifier > see slack!

File: tests/test_packages/test_connections/test_soef/models.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+#
+# ------------------------------------------------------------------------------
+
+"""This module contains models for soef connection tests."""
+
+from aea.helpers.search.models import Attribute, DataModel, Location
+
+
+AGENT_LOCATION_MODEL = DataModel(
+    "location_agent",
```
Comment (DavidMinarsch) [inline tests/test_packages/test_connections/test_soef/models.py:26]: see re naming above...also import constants for magic values from connection

File: packages/fetchai/connections/soef/connection.py (hunk 9 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py:317]: i dont know, new to protocol. probably it flushed with unregister., same for location.

File: packages/fetchai/connections/soef/connection.py (hunk 12 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py]: done

File: packages/fetchai/connections/soef/connection.py (hunk 10 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py]: updated

File: packages/fetchai/connections/soef/connection.py (hunk 11 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py]: done

File: packages/fetchai/connections/soef/connection.py (hunk 8 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py]: updated

File: packages/fetchai/connections/soef/connection.py (hunk 7 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py:273]: it's a primary code to perform every http request and return text, its used by generic oef request and in agent register

File: packages/fetchai/connections/soef/connection.py (hunk 6 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py]: done

File: packages/fetchai/connections/soef/connection.py (hunk 5 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py:221]: done

File: packages/fetchai/connections/soef/connection.py (hunk 4 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py:211]: ok

File: packages/fetchai/connections/soef/connection.py (hunk 3 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py:177]: done

File: packages/fetchai/connections/soef/connection.py (hunk 2 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py]: done

File: packages/fetchai/connections/soef/connection.py (hunk 1 above)
Comment (solarw) [reply packages/fetchai/connections/soef/connection.py]: done

**Pull request merged**